src/mesh.f90

```diff
@@ -0,0 +1,375 @@
+    logical, dimension(3), private :: periodic_dir ! Whether or not a direction has a periodic BC
+    integer, private :: sz
+    class(geo_t), allocatable :: geo ! object containing geometry information
+    class(parallel_t), allocatable :: par ! object containing parallel domain decomposition information
+  contains
+    procedure :: get_SZ
+
+    procedure :: get_n_groups_dir
+    procedure :: get_n_groups_phi
+    generic :: get_n_groups => get_n_groups_dir, get_n_groups_phi
+
+    procedure :: get_dims_dir
+    procedure :: get_dims_phi
+    procedure :: get_dims_phi_dataloc
+    generic :: get_dims => get_dims_dir, get_dims_phi, get_dims_phi_dataloc
+
+    procedure :: get_n_dir 
+    procedure :: get_n_phi
+    generic :: get_n => get_n_dir, get_n_phi
+
+    procedure :: get_padded_dims_phi
+    procedure :: get_padded_dims_dir
+    generic :: get_padded_dims => get_padded_dims_dir, get_padded_dims_phi
+  
+    procedure :: get_coordinates
+
+    procedure :: set_sz
+    procedure :: set_padded_dims
+  end type mesh_t
+
+  interface mesh_t
+    module procedure mesh_init
+  end interface mesh_t
+
+  contains 
+
+  function mesh_init(dims_global, nproc_dir, L_global) result(mesh)
+    !! Completely initialise the mesh object. 
+    !! Upon initialisation the mesh object can be read-only and shouldn't be edited
+    !! Takes as argument global information about the mesh like its length, number of cells and decomposition in each direction
+    integer, dimension(3), intent(in) :: dims_global
+    integer, dimension(3), intent(in) :: nproc_dir ! Number of proc in each direction
+    real(dp), dimension(3), intent(in) :: L_global
+    type(mesh_t) :: mesh
+
+    integer :: nx, ny, nz, dir
+    integer :: ierr
+    logical :: is_last_domain
+
+    allocate(mesh%geo)
+    allocate(mesh%par)
+    mesh%dims_global(:) = dims_global
+    ! Geometry
+    mesh%geo%L = L_global
+    mesh%geo%d = mesh%geo%L(:) / mesh%dims_global(:)
+
+    ! Parallel domain decomposition
+    mesh%par%nproc_dir(:) = nproc_dir
+    mesh%par%nproc = product(nproc_dir(:))
+    call MPI_Comm_rank(MPI_COMM_WORLD, mesh%par%nrank, ierr)
+    call MPI_Comm_size(MPI_COMM_WORLD, mesh%par%nproc, ierr)
+    call domain_decomposition(mesh)
+
+    ! Local mesh dimensions
+    nx = mesh%dims_global(1)/mesh%par%nproc_dir(1)
+    ny = mesh%dims_global(2)/mesh%par%nproc_dir(2)
+    nz = mesh%dims_global(3)/mesh%par%nproc_dir(3)
+
+
+    ! TODO: fixme: hard code no periodic boundary (for now)
+    mesh%periodic_dir(:) = .false.
+
+    ! Define number of cells and vertices in each direction
+    mesh%vert_dims = [nx, ny, nz]
+    mesh%cell_dims = mesh%vert_dims
+
+
+    do dir=1, 3
+      is_last_domain = (mesh%par%nrank_dir(dir) == mesh%par%nproc_dir(dir))
+      if (is_last_domain) then
+        if (mesh%periodic_dir(dir)) then
+          mesh%vert_dims(dir) = mesh%vert_dims(dir) + 1
```

Comment:
Vertex dims are always the dimensions we set in the simulation. Only the cell counts can be different. The number of cells in a direction is just one less than vertices in that direction if we're at the last region in the domain .and. the BC is .not. periodic. In all the other cases they're equal to each other.